python/pyiceberg/schema.py

```diff
@@ -638,3 +640,61 @@ def map(self, map_type: MapType, key_result: int, value_result: int) -> int:
 
     def primitive(self, primitive: PrimitiveType) -> int:
         return 0
+
+
+def assign_fresh_schema_ids(schema: Schema) -> Schema:
+    """Traverses the schema, and sets new IDs"""
+    schema_struct = visit(schema.as_struct(), _SetFreshIDs())
+
+    fresh_identifier_field_ids = []
+    new_schema = Schema(*schema_struct.fields)
+    for field_id in schema.identifier_field_ids:
+        original_field = schema.find_field(field_id)
+        if original_field is None:
+            raise ValueError(f"Could not find field: {field_id}")
+        fresh_field = new_schema.find_field(original_field.name)
+        if fresh_field is None:
+            raise ValueError(f"Could not lookup field in new schema: {original_field}")
+        fresh_identifier_field_ids.append(fresh_field.field_id)
+
+    return new_schema.copy(update={"identifier_field_ids": fresh_identifier_field_ids})
+
+
+class _SetFreshIDs(SchemaVisitor[IcebergType]):
```

Comment:
This is one area where we use pre-order rather than post-order in Java. It's not a problem as long as everything gets reassigned, but it is nice to get IDs assigned sequentially. That matches user expectations.

Definitely not a blocker. Up to you whether you want to add a `CustomOrderSchemaVisitor` for this purpose or not.

Reply:
Little effort and I think we need it anyway at some point. It is also nice to stick to the Java impl on this one.